### DISCUSSION TITLE
be_great v0.0.14

### DISCUSSION THREAD

**Discussion by unnir** [tabularis-ai/be_great#64] [Announcements]
**Mock data generation — no `fit()` required.** New `model.mock(schema=...)` builds plausible rows from a declarative schema instead of real training data. Per-column specs support `integer`, `precision`, `dist="normal"` with `mean`/`std`, categorical `weights`, and `null_prob` for realistic missing values. The existing `conditions=` operators work unchanged — schema declarations are implicit constraints, `conditions=` tighten them. Reproducible via `seed=`. Few-shot grounding via `examples=` (rows prepended as in-context demonstrations — big realism boost on LLMs that don't natively know the GReaT row format). Pairs well with [`tabularisai/Qwen3-0.3B-distil`](https://huggingface.co/tabularisai/Qwen3-0.3B-distil). Save/share schemas via `save_schema` / `load_schema` (JSON or YAML).

```python
from be_great import GReaT
from be_great.great_mock_datasets import save_schema, load_schema
 
schema = {
    "age":    {"type": "num", "range": (18, 99), "integer": True,
               "dist": "normal", "mean": 38, "std": 13, "null_prob": 0.05},
    "sex":    {"type": "cat", "values": ["Male", "Female", "Other"],
               "weights": [0.5, 0.45, 0.05]},
    "income": {"type": "cat", "values": ["<=50K", ">50K"], "weights": [0.76, 0.24]},
}
model = GReaT(llm="tabularisai/Qwen3-0.3B-distil")
mock = model.mock(schema=schema, n_samples=100, seed=42,
                  conditions={"age": ">= 40", "sex": "!= 'Male'"})
```

**chugchug as the default progress bar** for both training and sampling, replacing HF's tqdm. The new `ChugProgressCallback` shows a gradient bar with auto-colored metrics — loss goes green when improving, red when worsening — plus live `lr` and `epoch`. Notebooks get HTML/CSS gradient bars automatically. [`chugchug`](https://github.com/unnir/chugchug) `>= 0.1.3` is now a required dependency.

**Live synthesis-quality monitoring during training.** Pass `eval_data=` to `fit()` and, after each epoch (or every N steps via `eval_every=`), GReaT samples from the in-training model, scores it with `ColumnShapes`, and pushes `q=…` into the chugchug bar. A `🎯 [quality @ step N] column_shapes_mean = 0.XXXX` line is also printed. Watch synthesis quality climb in real time — training signal beyond raw loss.

**Auto device detection.** Generation methods now default to `device="auto"`: CUDA → MPS → CPU. Apple Silicon works out of the box. `dataloader_pin_memory` defaults to `False` on non-CUDA backends, silencing the MPS warning.

**First-token bias on constrained sampling.** When the schema declares `dist="normal"` or `weights`, the trie's first valid token is biased by the value-weight distribution — pulling generations toward the declared shape instead of uniform over the constraint set.

**Fix: garbage tails in guided sampling output** (`"Female, age is 73"`). Parser now picks the earliest delimiter among `;`, `,`, `\n` (plus `.` for non-numeric columns) instead of `;` first. Numeric columns preserve decimals.

minor: New `be_great/great_mock_datasets.py` cleanly separates the mock path from the fit path. Added `__repr__` to `GReaT`. Per-column `col_stats["precision"]` override in `_enumerate_numeric`. Version bumped `0.0.12` → `0.0.14` so source matches the PyPI release (`0.0.13` was tagged but never source-bumped).

<hr /><em>This discussion was created from the release <a href='https://github.com/tabularis-ai/be_great/releases/tag/'>be_great v0.0.14</a>.</em>
